R/lehdr.R

```diff
@@ -48,67 +51,105 @@ grab_lodes <- function(state, year, lodes_type = c("od", "rac", "wac"),
+  lodes_type <- match.arg(tolower(lodes_type), c("od", "rac", "wac"))
+  
+  # This handles the error in the `agg_geo` argument, if NULL it sets to block (the first non-null value)
+  agg_geo_to <- match.arg(tolower(agg_geo), c(NULL, "block", "bg", "tract", "county"))
   
+  # Setup col_types and url variables, used in the if statement
+  cols_types <- cols()
+  url <- as.character("")
+  
+  # Read LODES_types and set col_types
+  if (lodes_type == "od") {
+    url <- glue::glue("https://lehd.ces.census.gov/data/lodes/LODES7/{state}/{lodes_type}/{state}_{lodes_type}_{state_part}_{job_type}_{year}.csv.gz")
+    col_types <- cols(w_geocode = col_character(), 
+                      h_geocode = col_character(),
+                      createdate = col_character())
+  } else {
+    url <- glue::glue("https://lehd.ces.census.gov/data/lodes/LODES7/{state}/{lodes_type}/{state}_{lodes_type}_{segment}_{job_type}_{year}.csv.gz")    
+    # Avoid an error message by setting the col_type to NULL for those that won't appear
+    if (lodes_type == "rac") {
+      col_types <- cols(h_geocode = col_character(),
+                        createdate = col_character())
+    } else if (lodes_type == "wac") {
+      col_types <- cols(w_geocode = col_character(), 
+                        createdate = col_character())
+    }
+  }
+  
+  # On URL error, the likely culprit is the lack of state/year combination ...
   httr::stop_for_status(httr::HEAD(url),
-                        "data for this combination of features was not found on LODES")
+    paste("retrieve data for this combination of state and year on LODES.",
+          "Please see the most recent LEHD Technical Document for a list of available state/year.",
+          "https://lehd.ces.census.gov/data/lodes/LODES7/"
+    )
+  )
   
+  # Set download directory, check for cache
   download_dir <- path.expand(download_dir)
   if (!dir.exists(download_dir))
     dir.create(download_dir)
   fil <- file.path(download_dir, basename(url))
   
+  # If there is a cache, use it
   if(file.exists(fil)) {
     message(glue::glue("Cached version of file found in {fil}\n Reading now..."))
   } else {
     message(glue::glue("Downloading {url} to {fil} now..."))
     res <- httr::GET(url, httr::write_disk(fil))
   }
   
-  col_types <- cols(w_geocode = col_character(), 
-                    h_geocode = col_character(),
-                    createdate = col_character())
-  
-  if (lodes_type == "rac")
-    col_types$cols$w_geocode <- NULL
-  if (lodes_type == "wac")
-    col_types$cols$h_geocode <- NULL
-  
-  df <- suppressMessages(readr::read_csv(fil, col_types = col_types))
+  # Read in the data
+  lehdr_df <- suppressMessages(readr::read_csv(fil, col_types = col_types))
 
-  df <- df %>% mutate(
+  # Mutate the data based on year and state
+  lehdr_df <- lehdr_df %>% mutate(
     year=year,
-    State=toupper(state))
+    state=toupper(state))
   
-  if (tract==TRUE) {
```

Comment:
I abstracted this out in a function below so that it can be reused for block group and county aggregation.